subsys/nrf_rpc/nrf_rpc_ipc.c

```diff
@@ -0,0 +1,217 @@
+do {									       \
+	if ((len) > 32) {						       \
+		LOG_HEXDUMP_DBG(memory, 32, text " (truncated)");	       \
+	} else {							       \
+		LOG_HEXDUMP_DBG(memory, (len), text);			       \
+	}								       \
+} while (0)
+
+/* Translates error code from the lower layer to nRF RPC error code. */
+static int translate_error(int ll_err)
+{
+	switch (ll_err) {
+	case -EINVAL:
+		return -NRF_EINVAL;
+	case -EIO:
+		return -NRF_EIO;
+	case -EALREADY:
+		return -NRF_EALREADY;
+	case -EBADMSG:
+		return -NRF_EBADMSG;
+
+	case RPMSG_ERR_BUFF_SIZE:
+	case RPMSG_ERR_NO_MEM:
+	case RPMSG_ERR_NO_BUFF:
+		return -NRF_ENOMEM;
+	case RPMSG_ERR_PARAM:
+		return -NRF_EINVAL;
+	case RPMSG_ERR_DEV_STATE:
+	case RPMSG_ERR_INIT:
+	case RPMSG_ERR_ADDR:
+		return -NRF_EIO;
+
+	default:
+		if (ll_err < 0) {
+			return -NRF_EIO;
+		}
+		break;
+	}
+
+	return 0;
+}
+
+static void ept_bound(void *priv)
+{
+	const struct nrf_rpc_tr *transport = priv;
+	struct nrf_rpc_ipc *ipc_config = transport->ctx;
+
+	LOG_DBG("nRF RPC endpoint %s connected", ipc_config->endpoint.ept_cfg.name);
+
+	k_event_set(&ipc_config->endpoint.ept_bond, 0x01);
+}
+
+static void ept_received(const void *data, size_t len, void *priv)
+{
+	const struct nrf_rpc_tr *transport = priv;
+	struct nrf_rpc_ipc *ipc_config = transport->ctx;
+
+	__ASSERT_NO_MSG(data != NULL);
+	__ASSERT_NO_MSG(ipc_config->receive_cb != NULL);
+
+	DUMP_LIMITED_DBG(data, len, "Received");
+
+	ipc_config->receive_cb(transport, data, len, ipc_config->context);
+}
+
+static void ept_error(const char *message, void *priv)
+{
+	LOG_ERR("Endpoint error: \"%s\"", log_strdup(message));
+	__ASSERT_NO_MSG(false);
+}
+
+static int init(const struct nrf_rpc_tr *transport, nrf_rpc_tr_receive_handler_t receive_cb,
+		void *context)
+{
+	int err;
+	struct nrf_rpc_ipc *ipc_config = transport->ctx;
+	struct nrf_rpc_ipc_endpoint *endpoint = &ipc_config->endpoint;
+	struct ipc_ept_cfg *cfg = &ipc_config->endpoint.ept_cfg;
+
+	if (ipc_config->used) {
+		return 0;
+	}
+
+	if (receive_cb == NULL) {
+		LOG_ERR("No transport receive callback");
+		return -NRF_EINVAL;
+	}
+
+	err = ipc_service_open_instance(ipc_config->ipc);
+	if (err && err != -EALREADY) {
+		LOG_ERR("IPC service instance initialization failed: %d\n", err);
+		return translate_error(err);
+	}
+
+	ipc_config->receive_cb = receive_cb;
+	ipc_config->context = context;
+
+	cfg->cb.bound = ept_bound;
+	cfg->cb.received = ept_received;
+	cfg->cb.error = ept_error;
+	cfg->priv = (void *)transport;
+
+	k_event_init(&endpoint->ept_bond);
+
+	err = ipc_service_register_endpoint(ipc_config->ipc, &endpoint->ept, cfg);
+	if (err) {
+		LOG_ERR("Registering endpoint failed with %d", err);
+		return translate_error(err);
+	}
+
+	ipc_config->used = true;
+
+	if (!k_event_wait(&endpoint->ept_bond, 0x01, false,
+			  EPT_BIND_TIMEOUT)) {
+		ipc_config->used = false;
+
+		LOG_ERR("IPC endpoint bond timeout");
+		return translate_error(-NRF_EPIPE);
```

Comment:
It doesn't look like this error code needed translation (looks like native nRF RPC code).

Reply:
Good catch. Fixed